akka-actor/src/main/scala/akka/io/DirectByteBufferPool.scala

```diff
@@ -36,32 +36,32 @@ private[akka] class DirectByteBufferPool(defaultBufferSize: Int, maxPoolEntries:
   private def allocate(size: Int): ByteBuffer =
     ByteBuffer.allocateDirect(size)
 
-  @tailrec
-  private final def takeBufferFromPool(): ByteBuffer =
-    if (locked.compareAndSet(false, true)) {
-      val buffer =
-        try if (buffersInPool > 0) {
+  private final def takeBufferFromPool(): ByteBuffer = {
+    var buffer: ByteBuffer = null
+
+    if (lock.tryLock(1, TimeUnit.MILLISECONDS))
+      try
+        if (buffersInPool > 0) {
           buffersInPool -= 1
-          pool(buffersInPool)
-        } else null
-        finally locked.set(false)
+          buffer = pool(buffersInPool)
+        }
+      finally lock.unlock()
 
-      // allocate new and clear outside the lock
-      if (buffer == null)
-        allocate(defaultBufferSize)
-      else {
-        buffer.clear()
-        buffer
-      }
-    } else takeBufferFromPool() // spin while locked
+    // allocate new and clear outside the lock
+    if (buffer == null)
+      allocate(defaultBufferSize)
+    else {
+      buffer.clear()
+      buffer
+    }
+  }
 
-  @tailrec
   private final def offerBufferToPool(buf: ByteBuffer): Unit =
-    if (locked.compareAndSet(false, true))
-      try if (buffersInPool < maxPoolEntries) {
-        pool(buffersInPool) = buf
-        buffersInPool += 1
-      } // else let the buffer be gc'd
-      finally locked.set(false)
-    else offerBufferToPool(buf) // spin while locked
+    if (lock.tryLock(1, TimeUnit.MILLISECONDS))
+      try
+        if (buffersInPool < maxPoolEntries) {
+          pool(buffersInPool) = buf
+          buffersInPool += 1
+        } // else let the buffer be gc'd
```

Comment:
here we can use `sun.misc.Cleaner` as pointed out in #17194, but that is for another PR